src/renderers/dom/fiber/ReactDOMFiber.js

```diff
@@ -157,10 +157,19 @@ var ReactDOM = {
     }
     // Unsound duck typing.
     const component = (componentOrElement : any);
-    if (component.nodeType === 1) {
+    if (component.nodeType === 1) { // Element
       return component;
     }
-    return DOMRenderer.findHostInstance(component);
+    if (component.updater) { // Component
+      const whileRendering = component._reactInternalInstance.progressedChild;
+      if (!component.updater.isMounted(component) && whileRendering) {
```

Comment:
Ok, I tried to do that more efficiently, but I have hard time passing that information around without relying on optional parameters.